File: src/visual-graph/builder/nesting-kind-of.ts
import { SCOPE_TYPE } from "../../analyzer/scope-type.js";
import type { SerializedScope } from "../../ir/serialized/serialized-scope.js";
import { AST_TYPE } from "../../parser/ast-type.js";
import {
  NESTING_KIND,
  type NestingKind,
} from "../../serializer/nesting-kind.js";

export function nestingKindOf(scope: SerializedScope): NestingKind | null {
  if (scope.type === SCOPE_TYPE.Function) {
    return scope.functionExpressionScope ? null : NESTING_KIND.Function;
  }
  if (scope.type === SCOPE_TYPE.For) {
    return NESTING_KIND.For;
  }
  if (scope.type === SCOPE_TYPE.Switch) {
    return NESTING_KIND.Switch;
  }
  if (scope.type === SCOPE_TYPE.Catch) {
    return NESTING_KIND.TryCatchFinally;
  }
  if (scope.type === SCOPE_TYPE.Block) {
    const ctx = scope.blockContext;
    if (!ctx) {
      return NESTING_KIND.Block;
    }
    if (ctx.parentType === AST_TYPE.IfStatement) {
      return NESTING_KIND.If;
    }
    if (
      ctx.parentType === AST_TYPE.ForStatement ||
      ctx.parentType === AST_TYPE.ForInStatement ||
      ctx.parentType === AST_TYPE.ForOfStatement
    ) {
      return NESTING_KIND.For;
    }
    if (
      ctx.parentType === AST_TYPE.WhileStatement ||
      ctx.parentType === AST_TYPE.DoWhileStatement
    ) {
      return NESTING_KIND.While;
    }
    if (
      ctx.parentType === AST_TYPE.TryStatement ||
      ctx.parentType === AST_TYPE.CatchClause
    ) {
      return NESTING_KIND.TryCatchFinally;
    }
    if (ctx.parentType === AST_TYPE.SwitchStatement) {
      return NESTING_KIND.Switch;
    }
    return NESTING_KIND.Block;
  }
  return null;
}

File: crates/unsnarl-visual-graph/src/builder/nesting_kind_of.rs
//! Categorises a scope by the `NestingKind` it counts against —
//! the bucket the depth control machinery uses to decide whether
//! the scope collapses past the depth threshold.

use unsnarl_ir::nesting_kind::NestingKind;
use unsnarl_ir::scope::block_context::BlockContext;
use unsnarl_ir::scope_type::ScopeType;
use unsnarl_ir::serialized::SerializedScope;
use unsnarl_oxc_parity::AstType;

pub fn nesting_kind_of(scope: &SerializedScope) -> Option<NestingKind> {
    match scope.r#type {
        ScopeType::Function => {
            if scope.function_expression_scope {
                None
            } else {
                Some(NestingKind::Function)
            }
        }
        ScopeType::For => Some(NestingKind::For),
        ScopeType::Switch => Some(NestingKind::Switch),
        ScopeType::Catch => Some(NestingKind::TryCatchFinally),
        ScopeType::Block => {
            let Some(ctx) = scope.block_context.as_ref() else {
                return Some(NestingKind::Block);
            };
            Some(nesting_kind_from_block_context(ctx))
        }
        _ => None,
    }
}

fn nesting_kind_from_block_context(ctx: &BlockContext) -> NestingKind {
    match ctx.parent_type() {
        AstType::IfStatement => NestingKind::If,
        AstType::ForStatement | AstType::ForInStatement | AstType::ForOfStatement => {
            NestingKind::For
        }
        AstType::WhileStatement | AstType::DoWhileStatement => NestingKind::While,
        AstType::TryStatement | AstType::CatchClause => NestingKind::TryCatchFinally,
        AstType::SwitchStatement => NestingKind::Switch,
        _ => NestingKind::Block,
    }
}

#[cfg(test)]
#[path = "nesting_kind_of_test.rs"]
mod nesting_kind_of_test;
